### PR TITLE
dns: fix resolving STRICT_DNS cluster endpoint FQDN with trailing dot '.'

### PR DESCRIPTION
**Pull request by lmarszal**
Commit Message: dns: fix resolving STRICT_DNS cluster endpoint FQDN with trailing dot '.'
Additional Description: bump c-ares dependency to a version having https://github.com/c-ares/c-ares/commit/ba597817a51790a89020970da982283020312dbf issue resolved
Risk Level: Low
Testing: Manual test - repro steps from https://github.com/envoyproxy/envoy/issues/13587 now results in expected behaviour
Docs Changes: feature of adding trailing dot '.' to FQDN was never documented; not sure if applicable
Release Notes: N/A
Platform Specific Features: N/A

Fixes https://github.com/envoyproxy/envoy/issues/13587

### CODE REVIEW

**Comment by lmarszal** [comment]
CC @htuch

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -101,12 +101,12 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
-        version = "1.16.1",
-        sha256 = "d08312d0ecc3bd48eee0a4cc0d2137c9f194e0a28de2028928c0f6cae85f86ce",
+        version = "4d6975b3ec24aba44331d235bbcabf899be56cf2",
+        sha256 = "03f8470e6c6e39110557ff3d23d828c70267a1a057ef9b6e34215b6d4c7b8d6c",
         strip_prefix = "c-ares-{version}",
-        urls = ["https://github.com/c-ares/c-ares/releases/download/cares-{underscore_version}/c-ares-{version}.tar.gz"],
+        urls = ["https://github.com/c-ares/c-ares/archive/{version}.tar.gz"],
         use_category = ["dataplane_core", "controlplane"],
-        last_updated = "2020-05-11",
+        last_updated = "2020-10-19",
```
Comment (htuch) [inline bazel/repository_locations.bzl:109]: We're switching to make this release (commit) date. I think this is 2020-09-28 based on https://github.com/c-ares/c-ares/commit/4d6975b3ec24aba44331d235bbcabf899be56cf2.

File: bazel/foreign_cc/BUILD [hunk 2]
```diff
@@ -91,8 +91,8 @@ envoy_cmake_external(
         "//conditions:default": [],
     }),
     postfix_script = select({
-        "//bazel:windows_x86_64": "cp -L $EXT_BUILD_ROOT/external/com_github_c_ares_c_ares/nameser.h $INSTALLDIR/include/nameser.h",
-        "//conditions:default": "",
+        "//bazel:windows_x86_64": "cp -L $EXT_BUILD_ROOT/external/com_github_c_ares_c_ares/src/lib/nameser.h $INSTALLDIR/include/nameser.h && cp -L $EXT_BUILD_ROOT/external/com_github_c_ares_c_ares/src/lib/ares_dns.h $INSTALLDIR/include/ares_dns.h",
+        "//conditions:default": "cp -L $EXT_BUILD_ROOT/external/com_github_c_ares_c_ares/src/lib/ares_dns.h $INSTALLDIR/include/ares_dns.h",
```
Comment (htuch) [inline bazel/foreign_cc/BUILD:95]: Why is this now needed in non-Windows?

**Review by htuch — COMMENTED** [review]
Thanks for the fix, couple of short questions.

File: bazel/foreign_cc/BUILD (hunk 2 above)
Comment (lmarszal) [reply bazel/foreign_cc/BUILD:95]: There was a change in directory structure in `c-ares` (https://github.com/c-ares/c-ares/commit/0bf721cdd7f64b893c4ad015c341de680617411b). We're using macros from `ares_dns.h` in `dns_imp_test.cc` - this header file is no longer available after `make install`, so I had to make a "hack" in order to make it visible for us.

**Comment by lmarszal** [comment]
Run into some concurrency (tsan and asan) problems in upgraded `c-ares`:
```
WARNING: ThreadSanitizer: heap-use-after-free (pid=15)
  Read of size 1 at 0x7b04000018cf by main thread:
    #0 as_is_first /b/f/w/external/com_github_c_ares_c_ares/src/lib/ares_getaddrinfo.c:765:7 (dns_impl_test+0x2e658eb)
    #1 next_dns_lookup /b/f/w/external/com_github_c_ares_c_ares/src/lib/ares_getaddrinfo.c:693:11 (dns_impl_test+0x2e648b2)
    #2 next_lookup /b/f/w/external/com_github_c_ares_c_ares/src/lib/ares_getaddrinfo.c:513:15 (dns_impl_test+0x2e64714)
    #3 next_lookup /b/f/w/external/com_github_c_ares_c_ares/src/lib/ares_getaddrinfo.c:527:11 (dns_impl_test+0x2e647df)
    #4 ares_getaddrinfo /b/f/w/external/com_github_c_ares_c_ares/src/lib/ares_getaddrinfo.c:681:3 (dns_impl_test+0x2e63c22)
    #5 Envoy::Network::DnsResolverImpl::PendingResolution::getAddrInfo(int) /proc/self/cwd/source/common/network/dns_impl.cc:283:3 (dns_impl_test+0x2e457e4)
    #6 Envoy::Network::DnsResolverImpl::resolve(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Network::DnsLookupFamily, std::__1::function<void (Envoy::Network::DnsResolver::ResolutionStatus, std::__1::list<Envoy::Network::DnsResponse, std::__1::allocator<Envoy::Network::DnsResponse> >&&)>) /proc/self/cwd/source/common/network/dns_impl.cc:252:25 (dns_impl_test+0x2e46543)
    #7 Envoy::Network::DnsImplTest::resolveWithExpectations(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Network::DnsLookupFamily, Envoy::Network::DnsResolver::ResolutionStatus, std::__1::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::list<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::optional<std::__1::chrono::duration<long long, std::__1::ratio<1l, 1l> > >) /proc/self/cwd/test/common/network/dns_impl_test.cc:483:23 (dns_impl_test+0x1a7d0b1)
    #8 Envoy::Network::DnsImplTest_DestroyChannelOnRefused_Test::TestBody() /proc/self/cwd/test/common/network/dns_impl_test.cc:681:3 (dns_impl_test+0x1a63b27)
    #9 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10 (dns_impl_test+0x52da1cc)
    #10 void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2616:14 (dns_impl_test+0x52babbe)
    #11 testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2655:5 (dns_impl_test+0x529cbc1)
    #12 testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2832:11 (dns_impl_test+0x529da99)
    #13 testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28 (dns_impl_test+0x529e70e)
    #14 testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44 (dns_impl_test+0x52af03c)
    #15 bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10 (dns_impl_test+0x52e174c)
    #16 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2616:14 (dns_impl_test+0x52be9ee)
    #17 testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10 (dns_impl_test+0x52ae8cb)
    #18 RUN_ALL_TESTS() /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46 (dns_impl_test+0x2d06d27)
    #19 Envoy::TestRunner::RunTests(int, char**) /proc/self/cwd/test/test_runner.cc:155:10 (dns_impl_test+0x2d05d45)
    #20 main /proc/self/cwd/test/main.cc:34:10 (dns_impl_test+0x2d0319d)
```
I think I know where the problem is - I will update this PR after resolving this problem upstream.

**Comment by htuch** [comment]
Interesting, let us know if you need any points on Envoy's DNS implementation. I just merged https://github.com/envoyproxy/envoy/pull/13582, which makes the mentioned last_updated --> release_date rename, so you might want to merge master to pick that up. Thanks.

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (lmarszal) [reply bazel/repository_locations.bzl]: Updated branch from master (update last_updated -> release_date https://github.com/envoyproxy/envoy/pull/13644#issuecomment-713103944), fixed the date.

**Comment by lmarszal** [comment]
@htuch It's all green.

**Review by htuch — APPROVED** [review]
@lmarszal wow, thanks for the upstream fix! 👍

**Comment by daveey** [comment]
Thanks for fixing! We hit this problem when upgrading envoy from 1.14 -> 1.16. Will this change make it's way into a release any time soon?

**Comment by ramaraochavali** [comment]
@lizan @mattklein123 @PiotrSikora Is it possible to backport this to `1.16` release?

**Comment by ramaraochavali** [comment]
Discussed offline with @lizan . Since this has dependencies and hence an element of risk involved, backport for this is not recommended.

**Pull request merged**